File: 2020/03/Main.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;

public class Main {

    public static void main(String[] args) throws IOException {

        if(args.length != 1) {
            System.out.println("Usage: java -jar 3.jar [Datafile]");
            System.exit(0);
        }

        File dataFile = new File(args[0]);
        String content = readFile(dataFile);

        String[] lines = content.split("\n");

        long rightOneDownOne = traverseMap(1, 1, lines);

        long rightThreeDownOne = traverseMap(3, 1, lines);
        System.out.println("Part 1: " + rightThreeDownOne);

        long rightFiveDownOne = traverseMap(5, 1, lines);

        long rightSevenDownOne = traverseMap(7, 1, lines);

        long rightOneDownTwo = traverseMap(1, 2, lines);

        // Multiply all results together
        long product = 1;
        product *= rightOneDownOne;
        product *= rightThreeDownOne;
        product *= rightFiveDownOne;
        product *= rightSevenDownOne;
        product *= rightOneDownTwo;

        System.out.println("Part 2: " + product);
    }

    private static long traverseMap(int right, int down, String[] map) {
        // The position coordinates
        int x = 0;
        int y = 0;

        long encounteredTrees = 0;

        while (y < map.length) {
            if (x > map[0].length() - 1) {
                x = x - map[0].length();
            }

            char position = map[y].charAt(x);

            if(position == '#') {
                encounteredTrees++;
            }

            y += down;
            x += right;
        }

        return encounteredTrees;
    }

    private static String readFile(File file) throws IOException {
        if (!file.exists()) {
            System.out.println("The file could not be found. Please provide a valid path");
            System.exit(0);
        }

        StringBuilder builder = new StringBuilder();
        String line;
        BufferedReader reader = new BufferedReader(new FileReader(file));

        while((line = reader.readLine()) != null) {
            builder.append(line).append("\n");
        }

        return builder.toString();
    }

}

File: 2020/03/Main.cpp
# include<iostream>
# include <fstream>
# include <vector>


std::vector<std::string> readFile(std::string path);
unsigned long int traverseMap(int right, int down, std::vector<std::string> map);


int main(int argc, char* argv[]) {

    if(argc != 2) {
        std::cout << "Usage: ./Main [datafile]" << std::endl;
        return 1;
    }

    std::string path = argv[1];
    std::vector<std::string> lines = readFile(path);

    unsigned long int rightOneDownOne = traverseMap(1, 1, lines);
    unsigned long int rightThreeDownOne = traverseMap(3, 1, lines);
    unsigned long int rightFiveDownOne = traverseMap(5, 1, lines);
    unsigned long int rightSevenDownOne = traverseMap(7, 1, lines);
    unsigned long int rightOneDownTwo = traverseMap(1, 2, lines);

    unsigned long int product = 1;
    product *= rightOneDownOne;
    product *= rightThreeDownOne;
    product *= rightFiveDownOne;
    product *= rightSevenDownOne;
    product *= rightOneDownTwo;


    std::cout << "Part 1: " << rightThreeDownOne << std::endl;
    std::cout << "Part 2: " << product << std::endl;

}


std::vector<std::string> readFile(std::string path) {

    std::fstream file;
    file.open(path, std::ios::in);
    std::vector<std::string> content;

    if(file.is_open()) {
        std::string line;

        while(std::getline(file, line)) {
            content.push_back(line);
        }
        file.close();
    }


    return content;
}



unsigned long int traverseMap(int right, int down, std::vector<std::string> map) {
    unsigned int x = 0;
    unsigned int y = 0;

    unsigned long int encounteredTrees = 0;

    while(y < map.size()) {
        if(x > map[0].length() - 1) {
            x -= map[0].length();
        }

        char position = map[y][x];

        if(position == '#') {
            encounteredTrees++;
        }

        y += down;
        x += right;
    }

    return encounteredTrees;
}
